### ISSUE TITLE
fetchFromGitLab (etc) can't inhibit unpacking

### ISSUE THREAD

**Issue by evils** [NixOS/nixpkgs#255657]
### Describe the bug
it is currently not possible to prevent a source fetcher like `fetchFromGitLab` (or `fetchZip` and other derivatives) from unpacking the source

this currently prevents building  `kicad.libraries.packages3d` on hydra
because the unpacked source is massive, even though i've reduced the output size to below hydra's output limit
(~1gb download, ~5gb unpacked, 0.5gb output)

### Steps To Reproduce
1. build `kicad.libraries.packages3d` on hydra?
   1. see https://hydra.nixos.org/build/235419758 for such a failure
  
### Expected behavior
be able to do a `fetchFromGitLab` with `dontUnpack = true;` and let the `unpackPhase` in `mkDerivation` do the unpacking

### Notify maintainers
@Ericson2314  as a code owner of nixpkgs internals
@oxij as a recent active contributor to fetchZip (which is where i think this feature needs to be implemented)
@SuperSandro2000 as someone who added a bunch of features to fetchFromGitLab
none of you are probably the right people, but maybe one of you knows who to ping :)

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.12, NixOS, 23.11 (Tapir), 23.11.git.fbfdf7522eebM`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"musnix, nixos, nixos-hardware"`
 - channels(evils): `"home-manager, nixos-unstable, nixpkgs"`
 - nixpkgs: `/home/evils/.nix-defexpr/channels/nixpkgs`
```

**Comment by evils** [comment]
i'm playing with `fetchZip`'s `postFetch`, note that the hashing of the content has to happen before the repacking -_-
(unless the repacking is deterministic, which i would think isn't a guarantee
and would still mean i have to recreate the repacking stage in the prefetching of the url in assembling versions.nix...)

it seems to me the build directory's size is the limit triggering the hydra failure
(going by the failing stage being unpacking and not installing)

i guess that eliminates the options of just outputting the original tarball after checking its content
or even applying the compression that the cache probably does at some point, before the check

**Comment by oxij** [comment]
`tar` with the options above should be deterministic.

**Comment by evils** [comment]
this is starting to look like a feature request for a `repack` option for `fetchZip`

wouldn't this change need to be matched in `nix-prefetch-url`?

(seems like there's already issues about a mismatch between the two, https://github.com/NixOS/nixpkgs/issues/111508, https://github.com/NixOS/nix/issues/4499
maybe the solution is to share an implementation between `fetchZip` and `nix-prefetch-url`?
(which seems to be the approach taken by [nix-prefetch](https://github.com/msteen/nix-prefetch)))

**Comment by evils** [comment]
updated title and description to make this a feature request

(mostly as i can't seem to figure out how to implement it myself, if you know who could, please ping them)

**Comment by SuperSandro2000** [comment]
> mostly as i can't seem to figure out how to implement it myself, if you know who could, please ping them

We can't just run the following in postFetch:
- mv the files to a temporary directory
- zip them with the deterministic flags copied from somewhere in nixpkgs
- move the archive to $out

**Comment by evils** [comment]
seems we can

it's quite the mess though
```
name = if name == "packages3d" then "source.tar.gz" else "";
postFetch = if name == "packages3d" then ''
  mkdir -p $TMPDIR/repack
  mv $out/* $TMPDIR/repack/.
  rm -rf $out
  tar --create --file "$TMPDIR/repacked.tar.xz" --xz --sort=name --mode='a+rw' --mtime='1970-01-01' -C "$TMPDIR" repack
  mv $TMPDIR/repacked.tar.xz $out
''
```
(for testing i recommend changing `packages3d` to `templates` in the `name ==` as that's the smallest library)

apparently `unpackPhase` relies on the file extension to determine how to unpack things?

and this approach isn't compatible with prefetching the hash in my update script...

oh and it takes about 30 minutes on my ryzen 3600 machine with 32G RAM to compress the 5.4G output like this...

**Comment by SuperSandro2000** [comment]
> apparently `unpackPhase` relies on the file extension to determine how to unpack things?

I think so. This could be changed to read magic values but that would be a bit more complex and so far file extensions seems to be good enough.

> oh and it takes about 30 minutes on my ryzen 3600 machine with 32G RAM to compress the 5.4G output down to 379M like this...

I think this is required otherwise things will break when github changes it's archiving process or we prefetch an uploaded asset which can be downloaded with fetchurl.

**Comment by evils** [comment]
> I think this is required
i agree about the unpacking being required, 30 minutes of compression seems excessive

@oxij do you know if tar can use gzip deterministically?

**Comment by oxij** [comment]
I'm not sure. GZip header has file name and timestamp which are non-deterministic for our use case, but whether `tar` sets them, I don't know.

Run

```
drv=$(nix-instantiate -A package.src)
nix-store --realize --cores 0 $drv
nix-store --realize --cores 0 --check $drv
```

for your package to check for determinism first.

If it does not work you could do `tar --to-stdout --sort=name ... | gzip -n --best > output.tar.gz` instead.

**Comment by evils** [comment]
`gzip --best` is also quite slow, taking about 6 minutes to repack, vs less than 3 minutes for the stock compression level which still yields something around 950M

[here's my current implementation](https://github.com/NixOS/nixpkgs/commit/b019473fe0299ce2de6b7aafa8424bc0add4c757)

i still think this should be done in a repack option as it seems like i'm not the only one running into this
and i'd be calling the fetcher in my update script as well and prefer not to maintain the rather delicate bash in 2 places

and i feel like this could probably be integrated into the unpacking
(like: pipe the unpacking directly into repacking, maybe even use something like https://github.com/NixOS/nixpkgs/pull/214029 (gitlab has zip archives), as the download may be the biggest time bottleneck (it is here))
but i'd like input from someone more familiar with `fetchFromGitlab` / `fetchZip`

**Comment by SuperSandro2000** [comment]
We can't just ask upstream to upload a release which is not based on a git checkout? Then we could avoid this all together.

**Comment by evils** [comment]
we can ask upstream to do that
there is an [upstream issue](https://gitlab.com/kicad/libraries/kicad-packages3D/-/issues/369) about compressing the 3D models there and i've noted the utility to us

and i agree that that would be the most elegant solution
(note that it has some history associated with it, so i don't expect that to happen promptly)

but this issue is about nixpkgs not providing a good workaround for this kind of problem
so even if my specific case is resolved, i'll keep this issue open

**Comment by evils** [comment]
ping @emilylange as i just learned about `recompressTarball` in chrome (97738d7a1715d8022fd9d6e11e45586a56d9bdec)
which solves the same problem i have
and creates the same problem with breaking the update script...

**Comment by emilylange** [comment]
Yep and funnily enough, I've already been subscribed to this issue here :)

> and creates the same problem with breaking the update script...

For a few days, yes, because that was quite a hotfix. But no longer as of 5766d04f9662b948abadf2b6e589c0e9e1079681 (part of #262147).

`chromium`'s `recompressTarball`, however, is incredibly inefficient.

Sure, it uses multithreaded `zstd` which is much faster than single-threaded `xz`, but there is still so much redundant and time-consuming io, which is simply not needed if we would just pipe `curl` right into `xz` and `tar --exclude` (and then back into `tar`) instead.

I might try to implement something along those lines before the end of this year (assuming I find time for it).

Though, I am not sure if it would make sense to have something like that exported globally in the `pkgs` scope.

It's quite a balancing act, deciding whether the additional compute needed it worth the storage saving.

When `chromium`'s tarball exceeded `hydra.nixos.org`'s configured `max_output_size`[^1], we could have easily increased that limit instead.

In case of `chromium`, recompressing the tarball is very much is worth the additional compute. We use the same (recompressed) tarball to build `chromium` and `ungoogled-chromium` in unstable, stable [23.11] and, until old-stable [23.05] is EOL, old-stable [23.05] as well.

Note: `xz` decompression will become noticeably faster\*, when #263446 lands in `master`.

\*it depends, because multithreading is very much opt-in in `xz`. One can only decompress an `xz` tarball multithreaded, if that tarball was compressed multithreaded.

**Comment by SuperSandro2000** [comment]
> When `chromium`'s tarball exceeded `hydra.nixos.org`'s configured `max_output_size`[1](#user-content-fn-1-80cedb4b1ffa2a3f34d4d1ccfe4259ce), we could have easily increased that limit instead.

I think it is worth having that option. Bumping those limits probably hurts scalability. 

But we should add a good comment informing users about the drawbacks and not light hearted use it everywhere 3 days after the merge.

**Comment by evils** [comment]
> Though, I am not sure if it would make sense to have something like that exported globally in the pkgs scope.

my main motivation for wanting this functionality in fetchZIp or equivalent (i assume that's what you mean with "exported in the `pkgs` scope")
is to be able to use that implementation in my update script via nix-prefetch, instead of having to maintain the implementation in 2 places
(and to have a general solution to this kind of problem available)

> It's quite a balancing act, deciding whether the additional compute needed it worth the storage saving.

unless we're considering enabling it by default, i don't see much of a balancing act
it's not about saving storage, it's about having the source code available on hydra at all,
which seems to imply additional compute and storage elsewhere, at least in my case
or practically not being able to provide a functioning package, in chromium's case

increasing hydra's limit seems like more of a workaround than a solution to the problem of large intermediate step

**Comment by emilylange** [comment]
I think I understand now and agree that some sort of `recompress` option as part of `pkgs.fetchzip` would make sense for an extremely low number packages.

Unfortunately, I am unable to dedicate any time to this any time soon. I am just too busy.

**Comment by evils** [comment]
i've got a usable implementation at https://github.com/evils/nixpkgs/commit/749338bdf0fa5615ff4aa85eddb8c2fdfbbd59af
this lacks automatic setting of `name`, which when not adjusted results in `mkDerivation` failing to unpack the supplied source because it does not have the appropriate filename extension
(is this desired behaviour?)

this currently requires the user to manually set `name` at the place it's used, as here: https://github.com/evils/nixpkgs/commit/cc863366f6c935d2386a0b94dde7e760a9cf600b#diff-573832f7b46b69b2e997920b63f29adf9418a641ec3308589988b58d41d3b2bbR102
but considering this is not a feature we seem to want to encourage, maybe this is a desirable barrier?

i'm also not seeing a quick way to drop this in instead of the `recompressTarball` in `chromium`
and it does not seem to work for `fetchFromGitHub` in `clickhouse`, which also recompresses its output, but requires fetching via Git

**Comment by evils** [comment]
i've included an implementation in #290338
i feel like there's something wrong with including it in another PR,
but that PR already relies on it...

**Comment by evils** [comment]
switched to a separate PR for this, #291178

and after using `nix-prefetch` i've now realized a package local workaround for the output limit wouldn't have to be duplicated in the prefetcher as it would reuse it just the same as reusing the fetchZip one

but since i've got an apparently working general solution...
